Sources/PitchSpeller/Data Structures/_Graph.swift

```diff
@@ -0,0 +1,293 @@
+//
+//  _Graph.swift
+//  PitchSpeller
+//
+//  Created by Benjamin Wetherfield on 7/14/18.
+//
+
+// MARK: - Directedness Flags
+
+protocol Directedness { }
+protocol Directed: Directedness { }
+protocol Undirected: Directedness { }
+
+typealias DirectedOver = OrderedPair
+extension DirectedOver: Directed { }
+
+typealias UndirectedOver = UnorderedPair
+extension UndirectedOver: Undirected { }
+
+// MARK: - Weightedness Flags
+
+protocol Weightedness { }
+protocol Unweighted: Weightedness { }
+protocol AsWeight: Weightedness { }
+
+// Allows Double to be used as an edge weight
+extension Double: AsWeight { }
+enum WithoutWeights: Unweighted {
+    case unweighted
+}
+
+// Weightable, directable implementation of a Graph structure
+struct _Graph<Weight: Weightedness, Pair: SymmetricPair & Directedness & Hashable> where Pair.A: Hashable {
+    
+    // MARK: - Typealiases
+    
+    typealias Node = Pair.A
+    
+    struct Edge {
+        
+        // MARK: - Instance Properties
+        
+        let nodes: Pair
+        let weight: Weight
+        
+        // MARK: - Initializers
+        
+        init (_ a: _Graph.Node, _ b: _Graph.Node, withWeight weight: Weight) {
+            self.nodes = Pair(a, b)
+            self.weight = weight
+        }
+        
+        init (_ nodes: Pair, withWeight weight: Weight) {
+            self.nodes = nodes
+            self.weight = weight
+        }
+    }
+    
+    struct Path {
+        
+        // MARK: - Instance Properties
+        
+        let nodes: [Node]
+        let weights: [Pair: Weight]
+        
+        // MARK: - Initializers
+        
+        init (_ nodes: [Node], _ weights: [Pair: Weight]) {
+            self.nodes = nodes
+            self.weights = weights
+        }
+    }
+    
+    // MARK: - Instance Properties
+     
+    var nodes: Set<Node>
+    var adjacents: [Pair: Weight]
+    
+    var edges: [Edge] {
+        return adjacents.compactMap { Edge($0.key, withWeight: $0.value) }
```

Comment:
Does this work?

```Swift
return adjacents.compactMap(Edge.init)
```

Reply:
I think it would if I added an initializer that accepted (key, value) labeled tuple ?

Reply:
Okay, mind blown, it worked!